File: 162-find-peak-element/solution.ts
const INF = 2 ** 31 + 1;
function findPeakElement(nums: number[]): number {
  let start = 0;
  let end = nums.length - 1;
  while (start <= end) {
    const mid = Math.floor((start + end) / 2);
    if (
      nums[mid] > (nums[mid + 1] ?? -INF) &&
      nums[mid] > (nums[mid - 1] ?? -INF)
    ) {
      return mid;
    }
    if (nums[mid] > nums[mid + 1]) end = mid;
    else start = mid + 1;
  }
  return 0;
}

File: 162-find-peak-element/solution.cpp
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int n = nums.size(), l = 0, r = n - 1;
        while(l < r) {
            int m = l + (r - l) / 2;
            int prev = (m > 0) ? nums[m - 1] : INT_MIN;
            int next = (m < n - 1) ? nums[m + 1] : INT_MIN;
            if(prev < nums[m] && nums[m] > next) return m;
            if(prev > nums[m]) r = m;
            else l = m + 1;
        }
        return l;
    }
};
